Optimize FP8 performance, unified debug system & model caching fixes, 7B artifact fix

adrientoupet opened the pull request:
Hi @numz ,

I have done quite a lot of changes to the codebase - this is ahead of implementing GGUF.  I don't think I broke anything but I'm on leave and didn't have time to run extensive testing. If you can please test it with one of your generation before merging it, that would be amazing. See full breakdown below:

I'm doing this PR now otherwise it's going to be way too much to review with GGUF update as well.  Thank you sir!

----

## 🎯 Core Improvements

### 1. **Unified Debug System**
- Replaced all print statements with centralized `Debug` class featuring categorized logging, hierarchical timers, and memory tracking
- **Migrated `enable_debug` from BlockSwap node to main SeedVR2 node** for universal access to debugging
- BlockSwap operations now integrated with unified debug system instead of separate debugger

### 2. **Smart FP8 Handling** 
- **Critical change**: FP8 models now keep native FP8 storage, only converting to BFloat16 for arithmetic operations
- This makes FP8 models faster and more memory efficient than FP16 as intended
- Removed blanket model conversions, simplified FP8CompatibleDiT wrapper

### 3. **Model Registry & Multi-Repo Support**
- New registry system supporting multiple HuggingFace repos (numz/ and AInVFX/)
- Added mixed FP8 models to fix 7B artifacts
- Auto-discovery of user models in ComfyUI folder
- Infrastructure ready for GGUF quantized models

### 4. **Model Caching with Memory Leak Fixes**
- **Migrated `cache_model` from BlockSwap node to main SeedVR2 node** for broader availability
- Fixed critical memory leaks: proper RoPE cache cleanup, wrapper reference management, VAE intermediate tensor cleanup
- Partial vs full cleanup modes for efficient reuse

## 🔧 Technical Enhancements

### 5. **Code Organization**
- New `constants.py`, `model_registry.py`, `debug.py` modules
- Removed legacy code: unused compatibility hooks, `clear_rope_cache()`, dead VAE files
- Centralized path management and model discovery

### 6. **Performance & Memory**
- Better RoPE handling - modules moved to CPU before cleanup
- Added `torch.cuda.ipc_collect()` calls throughout for better GPU memory reclamation
- Debug logging added to existing optimizations for better visibility

### 7. **User Experience**
- Detailed parameter tooltips in ComfyUI node
- Better error messages with download instructions
- Visual logging categories with icons
- BlockSwap performance summaries

numz commented:
Impressive !!!
so much changes !!

I'll test it quickly, great thanks :)

FurkanGozukara commented:
yes there were a lot of ram leakage

right now i am struggling with 3b model vram leak too it used literally double vram compared to 7b model

7b model

```
🔍 PRE-SAMPLER: Set memory fraction to 100% for DiT inference
🔍 PRE-SAMPLER DEBUG: About to start EulerSampler
🔍 PRE-SAMPLER DEBUG: DiT device = cuda:0
🔍 PRE-SAMPLER DEBUG: VAE device = cuda:0
🔍 PRE-SAMPLER DEBUG: latents device = cuda:0, shape = torch.Size([102400, 16]), memory = 0.003 GB
🔍 PRE-SAMPLER DEBUG: latents_cond device = cuda:0, shape = torch.Size([102400, 17]), memory = 0.003 GB
🔍 PRE-SAMPLER DEBUG: text_pos_embeds device = cuda:0, shape = torch.Size([58, 5120]), memory = 0.001 GB
🔍 PRE-SAMPLER DEBUG: text_neg_embeds device = cuda:0, shape = torch.Size([64, 5120]), memory = 0.001 GB
🔍 PRE-SAMPLER DEBUG: Total tensor memory on GPU = 0.007 GB
🔍 PRE-SAMPLER DEBUG: Checking DiT blocks...
🔍 PRE-SAMPLER DEBUG: First DiT block device = cuda:0
🔍 SAMPLER DEBUG: Inside autocast, creating sampler lambda
🔍 SAMPLER DEBUG: self.dit device before sample = cuda:0
🔍 SAMPLER DEBUG: GPU memory allocated before sample call = 16.37 GB
🔍 EULER SAMPLER START: Entering sample method
🔍 EULER SAMPLER: x device = cuda:0, dtype = torch.bfloat16
🔍 DIT FORWARD: About to call self.dit() for positive prompt
🔍 DIT FORWARD: args.x_t device = cuda:0
🔍 DIT FORWARD: latents_cond device = cuda:0
🔍 DIT FORWARD: GPU memory before torch.cat = 16.37 GB
🔍 DIT FORWARD: GPU memory after torch.cat = 16.38 GB
🔍 DIT FORWARD: concat_result shape = torch.Size([102400, 33]), device = cuda:0
🔍 FP8CompatibleDiT FORWARD: First call to DiT forward
🔍 FP8CompatibleDiT FORWARD: Current device = cuda:0
🔍 FP8CompatibleDiT FORWARD: First block device = cuda:0
🔍 FP8CompatibleDiT FORWARD: Total blocks = 36
🔍 FP8CompatibleDiT: About to call self.dit_model forward
🔍 FP8CompatibleDiT: GPU memory before dit_model call = 16.39 GB
🔍 FP8CompatibleDiT: GPU memory after dit_model call = 16.40 GB
```

3b model

```
🔍 PRE-SAMPLER: Set memory fraction to 100% for DiT inference
🔍 PRE-SAMPLER DEBUG: About to start EulerSampler
🔍 PRE-SAMPLER DEBUG: DiT device = cuda:0
🔍 PRE-SAMPLER DEBUG: VAE device = cuda:0
🔍 PRE-SAMPLER DEBUG: latents device = cuda:0, shape = torch.Size([102400, 16]), memory = 0.003 GB
🔍 PRE-SAMPLER DEBUG: latents_cond device = cuda:0, shape = torch.Size([102400, 17]), memory = 0.003 GB
🔍 PRE-SAMPLER DEBUG: text_pos_embeds device = cuda:0, shape = torch.Size([58, 5120]), memory = 0.001 GB
🔍 PRE-SAMPLER DEBUG: text_neg_embeds device = cuda:0, shape = torch.Size([64, 5120]), memory = 0.001 GB
🔍 PRE-SAMPLER DEBUG: Total tensor memory on GPU = 0.007 GB
🔍 PRE-SAMPLER DEBUG: Checking DiT blocks...
🔍 PRE-SAMPLER DEBUG: First DiT block device = cuda:0
🔍 SAMPLER DEBUG: Inside autocast, creating sampler lambda
🔍 SAMPLER DEBUG: self.dit device before sample = cuda:0
🔍 SAMPLER DEBUG: GPU memory allocated before sample call = 7.36 GB
🔍 EULER SAMPLER START: Entering sample method
🔍 EULER SAMPLER: x device = cuda:0, dtype = torch.bfloat16
🔍 DIT FORWARD: About to call self.dit() for positive prompt
🔍 DIT FORWARD: args.x_t device = cuda:0
🔍 DIT FORWARD: latents_cond device = cuda:0
🔍 DIT FORWARD: GPU memory before torch.cat = 7.36 GB
🔍 DIT FORWARD: GPU memory after torch.cat = 7.37 GB
🔍 DIT FORWARD: concat_result shape = torch.Size([102400, 33]), device = cuda:0
🔍 FP8CompatibleDiT FORWARD: First call to DiT forward
🔍 FP8CompatibleDiT FORWARD: Current device = cuda:0
🔍 FP8CompatibleDiT FORWARD: First block device = cuda:0
🔍 FP8CompatibleDiT FORWARD: Total blocks = 32
🔍 FP8CompatibleDiT: About to call self.dit_model forward
🔍 FP8CompatibleDiT: GPU memory before dit_model call = 7.38 GB
🔍 FP8CompatibleDiT: GPU memory after dit_model call = 15.26 GB
```

FurkanGozukara commented:
@adrientoupet also it has bigger errors


here my run prompt

`python inference_cli.py --video_path "a.mp4" --model seedvr2_ema_7b_fp16.safetensors --resolution 1024 --output a_upscaled.mp4 --model_dir ./models --batch_size 5 --preserve_vram --cuda_device 1`

here logs

```
Microsoft Windows [Version 10.0.26100.4770]
(c) Microsoft Corporation. All rights reserved.

E:\Ultimate_Video_Processing_v1\venv\Scripts>activate

(venv) E:\Ultimate_Video_Processing_v1\venv\Scripts>cd..

(venv) E:\Ultimate_Video_Processing_v1\venv>cd..

(venv) E:\Ultimate_Video_Processing_v1>cd ComfyUI

(venv) E:\Ultimate_Video_Processing_v1\ComfyUI>cd custom_nodes

(venv) E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes>cd test2

(venv) E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2>python inference_cli.py --video_path "a.mp4" --model seedvr2_ema_7b_fp16.safetensors --resolution 1024 --output a_upscaled.mp4 --model_dir ./models --batch_size 5 --preserve_vram --cuda_device 1
🚀 SeedVR2 Video Upscaler CLI started at 2025-07-30 15:40:02
🎬 Extracting frames from video...
VRAM libre: 22.74 GB
🚀 Loading model_weight: 7b_fp16
🎯 Detected NaDiT 7B FP16
🧹 Clearing VRAM cache...
W0730 15:40:32.937000 28900 Lib\site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
🚀 Loading VAE SafeTensors: ./models\./ema_vae_fp16.safetensors

💡 TIP: For 20 frames, use batch_size=17 to avoid padding

🎬 Batch 1: frames 0-4
📹 Sequence of 5 frames
EulerSampler:   0%|                                                                              | 0/1 [00:00<?, ?it/s]   ⚠️ Flash Attention failed for blocks.0.attn.rope, using original: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument tensors in method wrapper_CUDA_cat)
   ⚠️ Flash Attention failed for blocks.0.attn, using original: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument tensors in method wrapper_CUDA_cat)
   ⚠️ Flash Attention failed for blocks.0.attn.rope, using original: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument tensors in method wrapper_CUDA_cat)
❌ Error in forward pass: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument tensors in method wrapper_CUDA_cat)
   Model type: NaDiT 7B=True, NaDiT v2 3B=True
   Args dtypes: []
   Kwargs dtypes: [('vid', torch.bfloat16), ('txt', torch.bfloat16), ('vid_shape', torch.int64), ('txt_shape', torch.int64), ('timestep', torch.float16)]
Process Process-2:
Traceback (most recent call last):
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 214, in flash_attention_forward
    return self._sdpa_attention_forward(original_forward, module, *args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 249, in _sdpa_attention_forward
    return original_forward(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 84, in forward
    freqs = cache("rope_freqs_3d", lambda: self.get_freqs(shape))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\common\cache.py", line 28, in __call__
    return fn()
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 84, in <lambda>
    freqs = cache("rope_freqs_3d", lambda: self.get_freqs(shape))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 102, in get_freqs
    return torch.cat(freq_list, dim=0)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument tensors in method wrapper_CUDA_cat)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 214, in flash_attention_forward
    return self._sdpa_attention_forward(original_forward, module, *args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 249, in _sdpa_attention_forward
    return original_forward(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\nablocks\mmsr_block.py", line 127, in forward
    vid_q, vid_k = self.rope(vid_q, vid_k, window_shape, cache_win)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1751, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1762, in _call_impl
    return forward_call(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 218, in flash_attention_forward
    return original_forward(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 84, in forward
    freqs = cache("rope_freqs_3d", lambda: self.get_freqs(shape))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\common\cache.py", line 28, in __call__
    return fn()
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 84, in <lambda>
    freqs = cache("rope_freqs_3d", lambda: self.get_freqs(shape))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 102, in get_freqs
    return torch.cat(freq_list, dim=0)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument tensors in method wrapper_CUDA_cat)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 214, in flash_attention_forward
    return self._sdpa_attention_forward(original_forward, module, *args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 249, in _sdpa_attention_forward
    return original_forward(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 84, in forward
    freqs = cache("rope_freqs_3d", lambda: self.get_freqs(shape))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\common\cache.py", line 28, in __call__
    return fn()
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 84, in <lambda>
    freqs = cache("rope_freqs_3d", lambda: self.get_freqs(shape))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 102, in get_freqs
    return torch.cat(freq_list, dim=0)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument tensors in method wrapper_CUDA_cat)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python310\lib\multiprocessing\process.py", line 314, in _bootstrap
    self.run()
  File "C:\Python310\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\inference_cli.py", line 238, in _worker_process
    result_tensor = generation_loop(
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\core\generation.py", line 393, in generation_loop
    samples = generation_step(runner, text_embeds, preserve_vram, cond_latents=cond_latents, temporal_overlap=temporal_overlap)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\core\generation.py", line 130, in generation_step
    video_tensors = runner.inference(
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\utils\_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\core\infer.py", line 380, in inference
    latents = self.sampler.sample(
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\common\diffusion\samplers\euler.py", line 84, in sample
    pred = f(SamplerModelArgs(x, t, i))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\core\infer.py", line 382, in <lambda>
    f=lambda args: classifier_free_guidance_dispatcher(
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\common\diffusion\utils.py", line 76, in classifier_free_guidance_dispatcher
    return pos()
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\core\infer.py", line 383, in <lambda>
    pos=lambda: self.dit(
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1751, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1762, in _call_impl
    return forward_call(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 402, in forward
    return self.dit_model(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1751, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1762, in _call_impl
    return forward_call(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\nadit.py", line 176, in forward
    vid, txt, vid_shape, txt_shape = gradient_checkpointing(
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\nadit.py", line 32, in gradient_checkpointing
    return module(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1751, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1762, in _call_impl
    return forward_call(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\nablocks\mmsr_block.py", line 238, in forward
    vid_attn, txt_attn = self.attn(vid_attn, txt_attn, vid_shape, txt_shape, cache)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1751, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1762, in _call_impl
    return forward_call(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 218, in flash_attention_forward
    return original_forward(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\nablocks\mmsr_block.py", line 127, in forward
    vid_q, vid_k = self.rope(vid_q, vid_k, window_shape, cache_win)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1751, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\venv\lib\site-packages\torch\nn\modules\module.py", line 1762, in _call_impl
    return forward_call(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\optimization\compatibility.py", line 218, in flash_attention_forward
    return original_forward(*args, **kwargs)
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 84, in forward
    freqs = cache("rope_freqs_3d", lambda: self.get_freqs(shape))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\common\cache.py", line 28, in __call__
    return fn()
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 84, in <lambda>
    freqs = cache("rope_freqs_3d", lambda: self.get_freqs(shape))
  File "E:\Ultimate_Video_Processing_v1\ComfyUI\custom_nodes\test2\src\models\dit\rope.py", line 102, in get_freqs
    return torch.cat(freq_list, dim=0)
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when checking argument for argument tensors in method wrapper_CUDA_cat)
EulerSampler:   0%|                                                                              | 0/1 [00:02<?, ?it/s]
```

adrientoupet commented:
@FurkanGozukara I'm confused about the issues you're reporting - if it's related to this PR please do it here. If it's related to another version, please check first if there another ticket already describing the issue, otherwise open a new one.

> right now i am struggling with 3b model vram leak too it used literally double vram compared to 7b model
I assume this is NOT related to this PR, correct?

>  your pull request missing this fix :
>   File "E:\Ultimate_Video_Processing_v1\test2\src\optimization\memory_manager.py", line 15, in <module>
>     from comfy import model_management as mm
> ModuleNotFoundError: No module named 'comfy' 
I'm surprised - it might be because of the conflict, depending how you sourced the files. Because if I check on my side, the fixes are there:

```
# git clone https://github.com/NumZ/ComfyUI-SeedVR2_VideoUpscaler.git
Cloning into 'ComfyUI-SeedVR2_VideoUpscaler'...
remote: Enumerating objects: 711, done.
remote: Counting objects: 100% (194/194), done.
remote: Compressing objects: 100% (37/37), done.
remote: Total 711 (delta 176), reused 157 (delta 157), pack-reused 517 (from 1)
Receiving objects: 100% (711/711), 3.21 MiB | 18.89 MiB/s, done.
Resolving deltas: 100% (415/415), done.
# cd ComfyUI-SeedVR2_VideoUpscaler/
# git init
Reinitialized existing Git repository in /data/ComfyUI-SeedVR2_VideoUpscaler/.git/
# git fetch https://github.com/numz/ComfyUI-SeedVR2_VideoUpscaler.git pull/92/head:models
remote: Enumerating objects: 80, done.
remote: Counting objects: 100% (31/31), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 80 (delta 28), reused 29 (delta 28), pack-reused 49 (from 1)
Unpacking objects: 100% (80/80), 60.05 KiB | 891.00 KiB/s, done.
From https://github.com/numz/ComfyUI-SeedVR2_VideoUpscaler
 * [new ref]         refs/pull/92/head -> models
# git switch models
Switched to branch 'models'
# cat src/optimization/memory_manager.py 
"""
Memory management module for SeedVR2
Handles VRAM usage, cache management, and memory optimization

Extracted from: seedvr2.py (lines 373-405, 607-626, 1016-1044)
"""

import os
import torch
import gc
import time
from typing import Tuple, Optional
from src.common.cache import Cache

try:
    from comfy import model_management as mm
    COMFYUI_AVAILABLE = True
except:
    COMFYUI_AVAILABLE = False
    pass
```

That said I found an issue related to that with the blockswap.py, so I'll correct it soon.

> also it has bigger errors
> here my run prompt

Good catch, thanks for reporting - I didn't test the update with the CLI script. Will look into that now.

> after spending huge time on vae vram leak only solution i found is deleting and recreating vae after each batch haha :D
This PR shouldn't suffer from that leak anymore. But if it does happy to look further into it. Thank you!

FurkanGozukara commented:
@adrientoupet ty i am testing cli. waiting your cli fix to test

JohnAlcatraz commented:
I also have some doubt if the fp8 vs fp16 for the model itself works correctly now - because I the same speed and VRAM usage no matter if I use the 7b fp16 or the 7b fp8 model. I need to look at the VRAM usage while I see the "Block X swap: ..." output in the log to judge if it's working correctly, right? And in a quick test with a short 25 frames video, I see both fp8 and fp16 use exactly the same 49% VRAM during those steps.

naxci1 commented:
It was running at 3.7 fps with the old version, I updated it today and the speed dropped to 3.36 fps. The model I'm using is 3b16, 16 GB VRAM.

JohnAlcatraz commented:
@adrientoupet Thanks for your detailed explanation!

But I'm still not sure if what I see regarding the VRAM usage makes sense... I'm testing with fp8 and a longer 81 frames video now. So I see this in the log:

```
🔀 BlockSwap memory configuration:
🔀   All 7994.62MB transformer blocks offloaded to cpu
🔀   VRAM usage: 0.00MB (blocks loaded on-demand during inference)
🔀   Total transformer blocks memory: 7994.62MB
🔀   I/O components offloaded: vid_in, txt_in, emb_in, vid_out
🔀   Non-blocking GPU transfers: Enabled
✅ BlockSwap configuration complete
```

So yeah, everything offloaded to the CPU. And by that point I see that in the log, the VRAM usage is ~10% or so, so almost nothing. So everything normal there.

But then I get to this point in the log:

```
🎬 ━━━ Batch 1/1: frames 0-80 ━━━
🎬 Video compute dtype: torch.bfloat16
📹 Sequence of 81 frames
🎨 VAE encoding precision: torch.bfloat16
ℹ️ Cond latents shape: torch.Size([21, 90, 156, 16])
```

And then the VRAM usage goes up to ~61%. Makes sense, it has to load the VAE to encode the frames.

But then after that finished, the VRAM usage should go to basically 0 again, right? Since then it's doing the block swaps:

```
🎯 Generation step precision: model=torch.float8_e4m3fn, inference=torch.bfloat16, autocast=torch.bfloat16
🎯 Model dtype: torch.float8_e4m3fn, using torch.bfloat16 for autocast
🎯 Text embeddings adapted to precision: torch.bfloat16
EulerSampler:   0%|                                                                                     | 0/1 [00:00<?, ?it/s]🔀 Io txt_in swap: 8.5ms
🔀 Io vid_in swap: 1.0ms
🔀 Io emb_in swap: 30.8ms
🔀 Block 0 swap: 263.4ms
🔀 Block 1 swap: 284.3ms
🔀 Block 2 swap: 230.5ms
```

But by that point, I still see the 61% VRAM usage. So it feels like either all blocks are in VRAM, or it incorrectly kept the VAE from the VAE encode in VRAM.

Then it finishes the block swaps, and does the VAE decode:

```
EulerSampler: 100%|█████████████████████████████████████████████████████████████████████████████| 1/1 [00:08<00:00,  8.51s/it]
🎯 VAE decode precision: torch.float16
ℹ️ Latents batch shape: torch.Size([1, 21, 90, 156, 16])
```

And by that point, it's up at 90% VRAM usage.

FurkanGozukara commented:
today i upscaled 121 frames at once and it is beyond amazing

however i used b200 since tiled vae we still lack - once tiled vae + block swapping unbeaatble 

https://github.com/user-attachments/assets/4c623441-6f98-46c2-9d2a-380ca44ea715

naxci1 commented:
@adrientoupet Ok, thank you. I'm looking forward to the new version. When will this be added to the main code and updated?

dimtoneff commented on src/interfaces/comfyui_node.py:
```diff
@@ -228,70 +272,88 @@ def cleanup(self, force_ram_cleanup: bool = True, keep_model_cached: bool = Fals
+
         return (sample,)
 
     def _progress_callback(self, batch_idx, total_batches, current_batch_frames, message=""):
```
"def __del__(self):" is calling self.cleanup with the wrong (old) arguments

naxci1 commented:
██████████████████████████████████████████████████████████████████████| 1/1 [00:06<00:00,  6.19s/it]
🎬 ━━━ Batch 14/15: frames 1313-1413 ━━━
📹 Sequence of 101 frames

In the other one, it showed the processing time in seconds. This was important to me. I wonder why you removed it? Can you bring it back?

kunibald413 commented:
@naxci1 ye i get that that's the merge request, i wondered which lines exactly achieve this

naxci1 commented:
> @kunibald413

<img width="924" height="616" alt="image" src="https://github.com/user-attachments/assets/02eaf2fc-0bed-49bf-adce-fb8b7a6133ba" />
<img width="1152" height="865" alt="image" src="https://github.com/user-attachments/assets/189c9478-5e02-4215-8417-ad6e96466f53" />

The modification date and title of the files are already written, and when you click on the file, the red part means it has been removed from the file, that is, deleted, and the green ones mean it has been added.

adrientoupet commented:
> In the other one, it showed the processing time in seconds. This was important to me. I wonder why you removed it? Can you bring it back?

Thank you @naxci1 for the feedback. All the timings are still available when you enable the debug mode on the SeedVR2 node. I removed it when debug is disabled to make the output less verbose as you're already getting the prompt execution time from ComfyUI.

e.g. debug off:

```
got prompt
🔄 Preparing model: seedvr2_ema_3b_fp8_e4m3fn.safetensors
💾 Cache miss: Creating new runner for model seedvr2_ema_3b_fp8_e4m3fn.safetensors
🚀 Creating DiT model on CPU for BlockSwap (will swap blocks to GPU during inference)
🚀 Loading model_weight: 3b_fp8
🎯 Detected NaDiT 3B FP8 - Converting RoPE freqs for FP8 compatibility
🎨 Loading VAE SafeTensors: /data/ComfyUI2505/models/SEEDVR2/./ema_vae_fp16.safetensors
🔀 BlockSwap configured: 36 blocks, non-blocking

🎬 Starting video upscaling generation...

🎬 ━━━ Batch 1/1: frames 0-0 ━━━
📹 Sequence of 1 frames
EulerSampler:   0%|                                                                      
EulerSampler: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:04<00:00,  4.76s/it]

ℹ️ Total frames: 1, shape per frame: 1200x1200x3
 
🎬 Video upscaling completed successfully!
Prompt executed in 40.20 seconds
```

debug on:
```
got prompt
🔄 Preparing model: seedvr2_ema_3b_fp8_e4m3fn.safetensors
💾 Cache miss: Creating new runner for model seedvr2_ema_3b_fp8_e4m3fn.safetensors
🚀 Creating DiT model on CPU for BlockSwap (will swap blocks to GPU during inference)
FETCH ComfyRegistry Data: 30/93
FETCH ComfyRegistry Data: 35/93
FETCH ComfyRegistry Data: 40/93
FETCH ComfyRegistry Data: 45/93
FETCH ComfyRegistry Data: 50/93
🚀 Loading model_weight: 3b_fp8
FETCH ComfyRegistry Data: 55/93
🎯 Detected NaDiT 3B FP8 - Converting RoPE freqs for FP8 compatibility
🎨 Loading VAE SafeTensors: /data/ComfyUI2505/models/SEEDVR2/./ema_vae_fp16.safetensors
🔀 BlockSwap configured: 36 blocks, non-blocking
FETCH ComfyRegistry Data: 60/93
 
🎬 Starting video upscaling generation...

(...)

─────────
⚡ Total execution: 36.39s
⚡   └─ Model preparation: 24.70s
⚡   └─ Video generation: 9.82s
⚡   └─ Final cleanup: 1.86s
 ─────────
Prompt executed in 37.08 seconds
```

Would that work for you or you think there is a valid reason to have it as well when debug is OFF  - and if so, which time were you interested about? Total execution ? or video generation time? And why? 

Thank you!

FurkanGozukara commented:
I am looking forward to tiled vae

This model needs it most due to vram requirement

naxci1 commented:
Thank you @adrientoupet for your answer. I disable debug because it slows down the speed. It's important for me to see how long each batch takes. I monitor its speed. If it's slow, I close it and reopen it, which speeds it up. Please display the processing time for each batch as before.

Regards.

<img width="1090" height="93" alt="image" src="https://github.com/user-attachments/assets/ca6b86a4-90ea-4f39-bc55-1aa13c6ef190" />

adrientoupet commented:
I'm closing this PR as it has been merged as part of #106 into the nightly branch. 

- @FurkanGozukara : you can now test the blockswap & vae tiling in the CLI tool
- @stepfunction83 : you can experiment with tiled vae in the ComfyUI node
- @naxci1 : I will do another update to get the performance closer to what it was before & look into bringing back the inference time for each batch with debug off
- @JohnAlcatraz: I will do another memory review/improved logging to ensure preserve_vram does the right thing as discussed above

Please keep it mind it's merged to nightly and the code/UI will likely change again before we push it to merge. Thank you!
